### PR TITLE
Start implementing new Comparer class

### PR DESCRIPTION
**Pull request by OrangeChannel**
Whew, have only tested that `stack_planes` and `tvbd_diff` still work as they did before (with some slight internal changes). The compare function likely will have to stay, and the code for `stack_compare` is pretty messy still so I'll push that later.

### CODE REVIEW

File: lvsfunc/comparison.py [hunk 1]
```diff
@@ -13,6 +17,316 @@
...
+        but specifying multiple clips in the **clips Dict with the same '' key will not work.
+        """
+        clips: List[vs.VideoNode] = []
+        for name, clip in self.clips.items():
+            if name.strip():
+                clips.append(clip.text.Text(text=name, alignment=self.label_alignment))
+            else:
+                clips.append(clip)
+
+        return clips
```
Comment (FichteFoll) [inline lvsfunc/comparison.py:116]:
```suggestion
        return [clip.text.Text(text=name, alignment=self.label_alignment)
                if name.strip()
                else clip
                for name, clip in self.clips.items()]
```

Comprehensions are nice.

Are VideoNodes hashable? If yes, return a set to reflect unorderedness.

File: lvsfunc/comparison.py [hunk 2]
```diff
@@ -13,6 +17,316 @@
...
+                                       Determines where to place clip name using ``vapoursynth.core.text.Text`` .
+                                       Will default to top-left for horizontal comparison or top-middle for vertical comparison
+                                       if not specified (Default: None)
+    :param clips:                      Keyword argument `name=clip` pairs for comparison. `name`s will be used to label the clips.
+                                       The order the `clips` are specified will be respected in the comparison.
+                                       If stacking horizontally, all clip heights must be the same.
+                                       If stacking vertically, all clip widths must be the same.
+    """
+
+    def __init__(self, direction: Direction,
```
Comment (FichteFoll) [inline lvsfunc/comparison.py:157]: If you added `direction` as a kw-only argument, you could use the same signature as Comparer.

File: lvsfunc/comparison.py [hunk 3]
```diff
@@ -13,6 +17,316 @@
...
+                            Determines where to place clip name using ``vapoursynth.core.text.Text`` (Default: 7)
+    :param arrangement:     2-dimension array (list of lists) of 0s and 1s representing a list of rows of clips(1)
+                            or blank spaces(0) (Default: None)
+    :param clips:           Keyword argument `name=clip` pairs for comparison. `name`s will be used to label the clips.
+                            The order the `clips` are specified will be respected in the comparison.
+                            The `width`, `height`, and `format` for all clips must be the same.
+    """
+
+    def __init__(self, label_alignment: Literal[1, 2, 3, 4, 5, 6, 7, 8, 9] = 7,
+                 arrangement: Optional[List[List[int]]] = None,
```
Comment (FichteFoll) [inline lvsfunc/comparison.py:244]: Same as for Stack's constructor signature.

File: lvsfunc/comparison.py [hunk 4]
```diff
@@ -13,6 +17,316 @@
...
+        else:
+            for i in range(1, num_blank_clips + 1):
+                array[-1][-i] = 0
+
+        self.arrangement: List[List[Union[int, vs.VideoNode]]] = array
+
+
+class Split(Stack):
+    """
+    Split and unlimited amount of clips into one VideoNode with the same dimensions as the original clips.
```
Comment (FichteFoll) [inline lvsfunc/comparison.py:306]: Typo "and" -> "an"

File: lvsfunc/comparison.py [hunk 5]
```diff
@@ -13,6 +17,316 @@
...
+    :param clips:           Keyword argument `name=clip` pairs for comparison. `name`s will be used to label the clips.
+                            The order the `clips` are specified will be respected in the comparison.
+                            The `width` and `height` for all clips must be the same.
+    """
+
+    def __init__(self, direction: Direction = Direction.HORIZONTAL,
+                 label_alignment: Optional[Literal[1, 2, 3, 4, 5, 6, 7, 8, 9]] = None,
+                 **clips: vs.VideoNode):
+        super().__init__(direction=direction, label_alignment=label_alignment, **clips)
+        self._smart_crop(direction=self.direction)
```
Comment (FichteFoll) [inline lvsfunc/comparison.py:327]: This function is only used here, so it should also be defined in the same class and not `Comparer`.

File: lvsfunc/comparison.py [hunk 6]
```diff
@@ -73,85 +387,91 @@ def _resample(clip: vs.VideoNode) -> vs.VideoNode:
...
-                  warn: bool = True) -> vs.VideoNode:
+                  make_diff: bool = True,
+                  height: int = 288,
+                  warn: Any
+                  ) -> vs.VideoNode:
     """
-    A simple wrapper that allows you to compare two clips by stacking them.
-    You can stack an infinite amount of clips.
+    Best to use when trying to match two sources frame-accurately.
+    Stacks two clips ontop of a MakeDiff clip for lining up clips from multiple sources.
```
Comment (FichteFoll) [inline lvsfunc/comparison.py:396]: The first sentence should always describe what a function is doing, not when it could be used.

File: lvsfunc/comparison.py [hunk 7]
```diff
@@ -73,85 +387,91 @@ def _resample(clip: vs.VideoNode) -> vs.VideoNode:
...
+    diff = core.std.MakeDiff(clips)
+    diff = diff.resize.Spline36(width=get_w(height, only_even=False) * 2, height=height * 2).text.FrameNum(8)
+    resized = [clips[0].resize.Spline36(width=get_w(height, only_even=False), height=height).text.Text('Clip A', 3),
+               clips[1].resize.Spline36(width=get_w(height, only_even=False), height=height).text.Text('Clip B', 1)]
 
-def stack_planes(clip: vs.VideoNode,
-                 stack_vertical: bool = False) -> vs.VideoNode:
+    return Stack(direction=Direction.VERTICAL,
+                 **{'': Stack(direction=Direction.HORIZONTAL,
+                              **{'': resized[0], ' ': resized[1]}
```
Comment (FichteFoll) [inline lvsfunc/comparison.py:424]: See how awkward this `**` stuff is? 😛 

In fact, seeing how it is used here, a list or generic iterable should also be accepted instead of only a dict.

File: lvsfunc/comparison.py [hunk 8]
```diff
@@ -169,42 +489,43 @@ def tvbd_diff(tv: vs.VideoNode, bd: vs.VideoNode,
...
-        stacked = core.std.StackVertical([core.std.StackHorizontal([tv, bd]), diff])
-        return core.std.Splice([stacked[f] for f in frames])
-
-
-# TODO: Write a comparison function that displays parts of clips side-by-side, similar to a slider.
-#       It should theoretically accept an infinite amount of clips
-#       and accurately split the width among all clips.
-#       Odd-resolution clips will also need to be taken into account.
+        diff = core.resize.Spline36(diff, get_w(height, only_even=False) * 2, height * 2).text.FrameNum(9)
+        tv, bd = tv.resize.Spline36(get_w(height, only_even=False), height).text.FrameNum(9), bd.resize.Spline36(get_w(height, only_even=False), height).text.FrameNum(9)
```
Comment (FichteFoll) [inline lvsfunc/comparison.py:522]:
```suggestion
        tv, bd = (c.resize.Spline36(get_w(height, only_even=False), height).text.FrameNum(9) for c in [tv, bd])
```

… or write it out in two lines so the code lines can be compared with the above `diff` easily.

File: lvsfunc/comparison.py [hunk 9]
```diff
@@ -13,6 +17,316 @@
...
+            self.width = None
+        if self.height == 0:
+            self.height = None  # this should be fixed on VapourSynth's side since variable-format uses None but width/height uses 0
+
+        self.formats: List[Union[None, str]] = []
+        for clip, name in self.clips:
+            try:
+                self.formats.append(clip.format.name)
+            except AttributeError:
+                self.formats.append(clip.format)
```
[the remark this answers is not in the document]
Comment (OrangeChannel) [reply lvsfunc/comparison.py:62]: Cleaned this up with getattr

File: lvsfunc/comparison.py (hunk 5 above)
Comment (OrangeChannel) [reply lvsfunc/comparison.py:327]: Originally was going to code a combination of Tile and Split that was also going to use this, but that's not happening for now so I moved it to Split.

File: lvsfunc/comparison.py [hunk 10]
```diff
@@ -27,111 +27,64 @@ class Direction(IntEnum):
...
 
-        elif direction == Direction.VERTICAL:
-            crop_height, overflow = self.height // self.num_clips, self.height % self.num_clips
+        self.num_clips = len(clips)
 
-            for key, (name, clip) in enumerate(self.clips.items()):
-                top_crop = crop_height * key
-                bottom_crop = crop_height * (self.num_clips - 1 - key)
+        widths: Set[int] = {clip.width for clip in self.clips}
+        self.width: Optional[int] = w if (w := widths.pop()) != 0 and len(widths) == 0 else None
```
Comment (FichteFoll) [inline lvsfunc/comparison.py:55]: You need to check `len(widths)` first and that should be `== 1`. Same for the following two attributes.

File: lvsfunc/comparison.py [hunk 11]
```diff
@@ -27,111 +27,64 @@ class Direction(IntEnum):
...
 
-    def compare(self) -> None:
+    def set_output(self) -> None:
         """
         Directly calls ``vapoursynth.VideoNode.set_output()`` on the :py:class:`lvsfunc.comparison.Comparer` generated comparison clip.
 
-        `Comparer(**kwargs, name1=clip1, name2=clip2, ...).compare()` is the intended use in encoding scripts linked with a previewer.
+        `Comparer(...).set_output()` is the intended use in encoding scripts linked with a previewer.
         """
         return self._compare().set_output()
```
Comment (FichteFoll) [inline lvsfunc/comparison.py]: This function seems pretty redundant to me when it's equivalent to `.clip.set_output()`.

File: lvsfunc/comparison.py [hunk 12]
```diff
@@ -143,60 +96,53 @@ class Stack(Comparer):
...
-                            The order the `clips` are specified will be respected in the comparison.
+    :param clips:
+    :param label_alignment:
     """
 
-    def __init__(self, label_alignment: Literal[1, 2, 3, 4, 5, 6, 7, 8, 9] = 7,
-                 **clips: vs.VideoNode):
-        super().__init__(label_alignment, **clips)
+    def __init__(self, clips: Union[Dict[str, vs.VideoNode], Sequence[vs.VideoNode]],
+                 /,
```
Comment (FichteFoll) [inline lvsfunc/comparison.py:137]: I think I prefer `/` as trailing the positional args in the first line, if feasible, but that's just my opinion.

File: lvsfunc/comparison.py [hunk 13]
```diff
@@ -27,111 +27,64 @@ class Direction(IntEnum):
...
-        clips: List[vs.VideoNode] = []
-        for name, clip in self.clips.items():
-            if name.strip():
-                clips.append(clip.text.Text(text=name, alignment=self.label_alignment))
-            else:
-                clips.append(clip)
-
-        return clips
+        if self.names:
+            self.clips = [clip.text.Text(text=name, alignment=self.label_alignment) if name.strip() else clip for clip, name in zip(self.clips, self.names)]
```
Comment (FichteFoll) [inline lvsfunc/comparison.py:68]: Why did you change this method to mutate `self.clips`?

File: lvsfunc/comparison.py [hunk 14]
```diff
@@ -231,31 +177,31 @@ class Tile(Comparer):
...
+        if arrangement is None:
             self._auto_arrangement()
+        else:
+            is_one_dim = len(arrangement) < 2 or all(len(row) == 1 for row in arrangement)
+            if is_one_dim:
+                raise ValueError("Tile: use Stack instead if the array is one dimensional")
+            self.arrangement: List[List[int]] = arrangement
 
-        self.BlankClip: vs.VideoNode = core.std.BlankClip(clip=list(self.clips.values())[0], keep=1)
+        self.BlankClip = core.std.BlankClip(clip=self.clips[0], keep=1)
```
Comment (FichteFoll) [inline lvsfunc/comparison.py:204]: Why is this attribute TitleCased?

File: lvsfunc/comparison.py [hunk 15]
```diff
@@ -274,12 +220,14 @@ def __init__(self, label_alignment: Literal[1, 2, 3, 4, 5, 6, 7, 8, 9] = 7,
...
+        self._markclips()
+        clip_arrangement: List[List[Union[vs.VideoNode]]] = []
         for row_key in range(len(self.arrangement)):
+            clip_arrangement.append([])
             for key, elem in enumerate(self.arrangement[row_key]):
-                self.arrangement[row_key][key] = clips.pop(0) if elem else self.BlankClip
+                clip_arrangement[row_key][key] = self.clips.pop(0) if elem else self.BlankClip
 
-        rows: List[vs.VideoNode] = [core.std.StackHorizontal(row) for row in self.arrangement]
+        rows = [core.std.StackHorizontal(row) for row in clip_arrangement]
```
Comment (FichteFoll) [inline lvsfunc/comparison.py]: Unless I am mistaken, this simplifies nicely to
```py
        rows = [core.std.StackHorizontal([self.clips.pop(0) if elem else self.BlankClip for elem in row])
                for row in self.arrangement]
```

File: lvsfunc/comparison.py [hunk 16]
```diff
@@ -298,33 +246,68 @@ def _auto_arrangement(self):
...
+                left_crop = crop_width * key
+                right_crop = crop_width * (self.num_clips - 1 - key)
+
+                if key != (self.num_clips - 1):
+                    right_crop += overflow
+
+                self.clips[key] = clip.std.Crop(left=left_crop, right=right_crop)
+
+        elif self.direction == Direction.VERTICAL:
+            crop_height, overflow = self.height // self.num_clips, self.height % self.num_clips
```
Comment (FichteFoll) [inline lvsfunc/comparison.py:301]: Check out `divmod`.

File: lvsfunc/comparison.py (hunk 13 above)
Comment (OrangeChannel) [reply lvsfunc/comparison.py:68]: I think originally I was using this to return a list of clips as I was essentially `list()`-ing an items view. There's no point in it returning anything now since clips is always a list.

File: lvsfunc/comparison.py (hunk 12 above)
Comment (petzku) [reply lvsfunc/comparison.py:137]: In my opinion, it's clearer to have `/` on a line of its own when the args are already split over multiple lines. It gives a clear split of the "sections" at a glance, although that might not be necessary here as all sections individually fit in a single line.

File: lvsfunc/comparison.py (hunk 11 above)
Comment (OrangeChannel) [reply lvsfunc/comparison.py]: Script writers are going to be used to doing `compare(clipa, clipb).set_output()` (probably) and this seemed like a nice convenient way of still being able to do that without using the clip return in-between.

File: lvsfunc/comparison.py (hunk 10 above)
Comment (OrangeChannel) [reply lvsfunc/comparison.py:55]: For `format` yes, that was a typo, but for these I'm pretty sure this is correct even if messy. The assignment operator (the pop) happens first, so the length should become `0` if there's only one width in the set. Tested with Stack and this seems to work how I intended. If there's a cleaner way of doing this on one line I'd change it since I didn't really give it much thought.

File: lvsfunc/comparison.py (hunk 2 above)
Comment (OrangeChannel) [reply lvsfunc/comparison.py]: Can you explain this?

File: lvsfunc/comparison.py (hunk 10 above)
Comment (FichteFoll) [reply lvsfunc/comparison.py:55]: > The assignment operator (the pop) happens first, so the length should become `0` if there's only one width in the set.

Ah, my thought was that pop could fail if `width` is empty, but that could only happen if `self.clips` was empty, which is bad usage in the first place. Regardless, I believe checking the length first and *then* checking the value of the only entry is more true to what happens.

File: lvsfunc/comparison.py (hunk 11 above)
Comment (FichteFoll) [reply lvsfunc/comparison.py]: Seems a bit dishonest to kind of pretend you have a clip but only implement `set_output()`. If the usage of `compare` is important, then that should just return the Comparer's `clip` immediately.

File: lvsfunc/comparison.py (hunk 12 above)
Comment (FichteFoll) [reply lvsfunc/comparison.py:137]: I would agree, if `clips` was not on the same line as `self` :wink:

File: lvsfunc/comparison.py (hunk 13 above)
Comment (FichteFoll) [reply lvsfunc/comparison.py:68]: From experience, dealing with state on an instance when the state is only used at one point in the pipeline (in `.compare`), you're 

1. pretending that there is internal state that persists across runs, which it does not,
2. causing inconsistencies of state to become more likely.

Unless I misremember things from the aruably really quick review from last night.

File: lvsfunc/comparison.py (hunk 10 above)
Comment (OrangeChannel) [reply lvsfunc/comparison.py:55]: I have a check to make sure len(clips) >= 2 so self.clips can't be empty.

File: lvsfunc/comparison.py (hunk 13 above)
Comment (OrangeChannel) [reply lvsfunc/comparison.py:73]: So I should change it back to how it was originally? I only understand a portion of your comment but don't see how it applies here.

Comment (FichteFoll) [reply lvsfunc/comparison.py:73]: Yes. I'll try to explain it differently:

The implemented classes (following the `Comparer` interface) currently do the following:
1. take a few parameters as input
2. produce a clip as a direct result of the input

In a sense, they are functions with inputs and a single output. This is a linear process.

They are implemented as classes to make use of polymorphism, which is a reasonable use case. Since they are a class, the inputs and some intermediate states used to calculate the output can also be accessed by callers (i.e. `self.width`), but that's more of an implementation through polymorphism. However, `self.clips` is mutated, which means even though it *could* be used to refer to the input variables, this is only true until the first access of `clip` (which is a property and does not indicate mutation of the class's state), after which `self.clips` refers to whatever `_markclips` did with them. Additionally, this intermediate state isn't cached or checked within the implementation itself, so accessing `clip` twice will reapply `_markclips`.

There are a few solutions to this:

1. Make `_markclips` return the modified clips instead of writing to `self.clips`.
2. Cache the result of `clip` with `functools.cached_property`. This should be documented.
3. Change `clip` into a classmethod that forwards its args to the constructor and then returns the result of its `_compare` (which might as well be inlined).
4. Implement the comparers as classmethods of `Comparer` and have them used as `Comparer.stack(...)` or `Comparer.interleave(...)`, returning the resulting clip.

All of these have their advantages and some only address part of the issues I mentioned, while others cover them up. In the end, I think what influences the decision the most is the answer to the question "What should instances of `Comparer` be used for?".

Comment (OrangeChannel) [reply lvsfunc/comparison.py]: Addressed. Split's `_smart_crop()` *has* to mutate `self.clips` though, I tried changing it and it would require re-writing it so much that there would be no point of it inheriting `Stack`, nor `Comparer`.

**Comment by OrangeChannel** [comment]
Once @FichteFoll checks the recent commits, will re-base on master, and actually break this into just one or two logical commits, probably one for implementing the comparers and one for changing the current functions to use them.

File: lvsfunc/comparison.py (hunk 10 above)
Comment (OrangeChannel) [reply lvsfunc/comparison.py:61]: The way it works currently is to prevent a "w might be referenced before assignment" error because checking the list length first and then popping *might not* pop the value if the length bool is false.

Comment (FichteFoll) [reply lvsfunc/comparison.py:61]: Sure, but then the expression wouldn't evaluate to true and the `if-true` expression isn't evaluated. In reality, it doesn't make a difference here, but it would if `widths` was empty.

Just so it's clear what I'm talking about:
```suggestion
        self.width: Optional[int] = w if len(widths) == 1 and (w := widths.pop()) != 0 else None
```

File: lvsfunc/comparison.py [hunk 17]
```diff
@@ -164,31 +164,31 @@ class Tile(Comparer):  # TODO: check this documentation will read correctly on t
...
-    For custom arrangements, such as (for 4 clips):
-    [
-     [0, 1, 0, 1],
-     [1],
-     [0, 1]
-    ],
-    the rows will be auto-padded with `0` to be the same length.
+    where `n` is equivalent to ``math.ceil(math.sqrt(len(clips))``\ .
+
+    | For example, for 3 clips, the automatic arrangement becomes:
```
Comment (FichteFoll) [inline lvsfunc/comparison.py:169]: In Sphinx, you can start a code block by ending the previous paragprah with `::` and indenting the following text by … probaby 4 spaces?

File: lvsfunc/comparison.py [hunk 18]
```diff
@@ -65,12 +65,14 @@ def __init__(self, clips: Union[Dict[str, vs.VideoNode], Sequence[vs.VideoNode]]
...
         self.format: Optional[str] = formats.pop() if len(formats) == 1 else None
 
-    def _markclips(self) -> None:
+    def _markclips(self) -> List[vs.VideoNode]:
         """
         If a `name` is only space characters, `'   '`, for example, the name will not be overlayed on the clip.
         """
         if self.names:
-            self.clips = [clip.text.Text(text=name, alignment=self.label_alignment) if name.strip() else clip for clip, name in zip(self.clips, self.names)]
+            return [clip.text.Text(text=name, alignment=self.label_alignment) if name.strip() else clip for clip, name in zip(self.clips, self.names)]
```
[the remark this answers is not in the document]
Comment (OrangeChannel) [reply lvsfunc/comparison.py:74]: The linter uses 120 but I don't know if this PR will get the actions done on it since I started it before the linter/flake8 branch was merged.

File: lvsfunc/comparison.py (hunk 17 above)
Comment (OrangeChannel) [reply lvsfunc/comparison.py:169]: I'll need louis to test building the docs but yea, this seems like a better idea.

Comment (heicrd) [reply lvsfunc/comparison.py:169]: `python setup.py build_sphinx` then open `build/sphinx/html/inde.html` in a browser

File: lvsfunc/comparison.py [hunk 19]
```diff
@@ -2,17 +2,347 @@
...
+            self.arrangement = arrangement
+
+        self.blank_clip = core.std.BlankClip(clip=self.clips[0], keep=1)
+
+        if not all(len(row) == (max_length := max(len(row) for row in self.arrangement)) for row in self.arrangement):
+            for row in self.arrangement:
+                if len(row) != max_length:
+                    diff = max_length - len(row)
+                    for _ in range(diff):
+                        row.append(0)  # padding all rows to the same length
```
Comment (FichteFoll) [inline lvsfunc/comparison.py:239]:
```suggestion
        max_length = max(self.arrangement, key=len)
        for row in self.arrangement:
            if len(row) != max_length:
                row.extend([0] * (max_length - len(row)))
```

File: lvsfunc/comparison.py [hunk 20]
```diff
@@ -2,17 +2,347 @@
...
+
+        array_count = 0
+        for list_ in self.arrangement:
+            for elem in list_:
+                if elem:
+                    array_count += 1
```
Comment (FichteFoll) [inline lvsfunc/comparison.py:245]:
```suggestion
        array_count = sum(map(sum, self.arrangement))
```
since all cells are either `0` or `1`.

File: lvsfunc/comparison.py [hunk 21]
```diff
@@ -2,17 +2,347 @@
...
+
+        if num_blank_clips >= dimension:
+            array.pop(-1)
+            for i in range(1, num_blank_clips - dimension + 1):
+                array[-1][-i] = 0
+        else:
+            for i in range(1, num_blank_clips + 1):
+                array[-1][-i] = 0
+
+        return array
```
Comment (FichteFoll) [inline lvsfunc/comparison.py:271]:
```suggestion
    def _auto_arrangement(self) -> List[List[int]]:
        dimension = 1 + math.isqrt(self.num_clips - 1)
        cells = [1] * self.num_clips + [0] * (dimension ** 2 - self.num_clips)
        return list(map(list, grouper(array, dimension)))
```

grouper is an itertools recipe. Can be either copied to the file or imported from `more-itertools`.

**Comment by OrangeChannel** [comment]
All classes/functions tested and working. Documentation also tested and looks good. Ready for merge if there's no more code suggestions.

File: lvsfunc/comparison.py [hunk 22]
```diff
@@ -169,44 +170,44 @@ class Tile(Comparer):
 
     The arrangement of the clips can be specified with the `arrangement` parameter.
     Rows are specified as lists of ints inside of a larger list specifying the order of the rows.
-    Think of this as a 2-dimensional array of `0` or `1` with `0` representing an empty slot and `1` representing the
+    Think of this as a 2-dimensional array of 0s and 1s with `0` representing an empty slot and `1` representing the
     next clip in the sequence.
 
     If `arrangement` is not specified, the function will attempt to fill a square with dimensions `n x n`
-    where `n` is equivalent to ``math.ceil(math.sqrt(len(clips))``.::
+    where `n` is equivalent to ``math.ceil(math.sqrt(len(clips))``. The bottom rows will be dropped if empty. ::
```
Comment (FichteFoll) [inline lvsfunc/comparison.py:179]: I'm just noticing this, but the `::` at the end are meant to be use as a replacement of a single `:` followed by a code block in the next paragraph., e.g. `Examples::\n\n    code here`.

What you would want here is the [code-block directive](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-code-block)

File: lvsfunc/comparison.py [hunk 23]
```diff
@@ -594,21 +579,21 @@ def stack_vertical(*clips: vs.VideoNode) -> vs.VideoNode:
 
     :param clips: Clips for comparison (order is kept top to bottom)
 
-    :return: Returns a vertical stack of the clips
+    :return: Returns a vertical stack of the `clips`
     """
     return Stack(clips, direction=Direction.VERTICAL).clip
 
 
 def tile(**clips: vs.VideoNode) -> vs.VideoNode:
```
Comment (FichteFoll) [inline lvsfunc/comparison.py:593]: I just now noticed that these convenience functions also/still use `*` and `**` instead of `clips: Union[Dict[str, vs.VideoNode], Sequence[vs.VideoNode]]`, which is a waste imo. Unless backwards-compatibility is important (not sure if even applicable), you should consider changing these to what `Comparer` uses in the constructor, since you're just forwarding that anyway. (You could use a type alias for this since you'll be using it quite a few times, I suppose.

Compare these:
```py
split = lvf.comparison.split(v1=src_a, v2=src_b, v1_again=src_a, v2_again=src_b, v1_again_again=src_a)
# vs
split = lvf.comparison.split([src_a, src_b, src_a, src_b, src_a])
```

If it would break compatibility and that is deemed important … maybe deprecate them?

File: lvsfunc/comparison.py [hunk 24]
```diff
@@ -247,12 +249,12 @@ def _compare(self) -> vs.VideoNode:
...
 
     def _auto_arrangement(self) -> List[List[int]]:
-        def _grouper(iterable, n, fillvalue=None):  # type: ignore
+        def _grouper(iterable: Iterable[T], n: int, fillvalue: Optional[T] = None) -> Iterator[Tuple[T, ...]]:
             args = [iter(iterable)] * n
             return zip_longest(*args, fillvalue=fillvalue)
 
         dimension = 1 + math.isqrt(self.num_clips - 1)
-        return list(map(list, _grouper([1] * self.num_clips, dimension, 0)))
+        return list(map(lambda x: list(x), _grouper([1] * self.num_clips, dimension, 0)))
```
Comment (FichteFoll) [inline lvsfunc/comparison.py:257]: if `map(list, …)` doesn't work for some reason, better use a comprehension. You'll almost never want to use a lambda inside a map.

```suggestion
        return [list(x) for x in _grouper([1] * self.num_clips, dimension, 0)]
```

Also note that for a clip size of 20, for example, you'll end up with a dimenion of 5 but because 20 is divided by 5, the resulting clip's aspect ratio will change. I thought this was unintended, which is why I wrote my suggestion differently (and without `fillvalue`).

File: lvsfunc/comparison.py (hunk 22 above)
Comment (OrangeChannel) [reply lvsfunc/comparison.py:179]: https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#literal-blocks

It says if it's preceded by whitespace it's removed, which it is in the docs I built when I was testing this. Not sure how to use the code-block directive in the code, that seems like something that would go into the .rst file, no?

Comment (FichteFoll) [reply lvsfunc/comparison.py:179]: > If it is preceded by whitespace, the marker is removed.

Oh, I didn't know about that. If it works, then that's fine, but you could still try the directive (with code highlighting). Iirc, reST is interpreted literally in docstrings, but I could be wrong as it's been a while for me.

File: lvsfunc/comparison.py (hunk 24 above)
Comment (OrangeChannel) [reply lvsfunc/comparison.py:257]: I don't think the aspect ratio was meant to be guaranteed to stay the same, maybe I could put that as an option? But realistically you can specify your own square arrangement if you'd want to do that. As for the list comprehension, yea that seems a bit nicer to read.

File: lvsfunc/comparison.py (hunk 22 above)
Comment (OrangeChannel) [reply lvsfunc/comparison.py:179]: The way it works currently seems to asssume python for the code block syntax highlighting, which works in this case since the list format would be how you would specify in python.

File: lvsfunc/comparison.py (hunk 24 above)
Comment (OrangeChannel) [reply lvsfunc/comparison.py]: Actually, I quite like the way it handles dropping blank rows now in the auto_arrangement because it results in the most optimal (spacial-wise and filesize if you were to export it as an image) way of tiling the clips. I don't think anyone would be encoding something like this as this would be *very* ram intensive and not sure what resolutions common encoders even support.

File: lvsfunc/comparison.py (hunk 23 above)
Comment (OrangeChannel) [reply lvsfunc/comparison.py:593]: I did this intentionally, because most (or at least the ones I know of) comparison functions or functions dealing with multiple clips usually work like this: `func(clipa, clipb, *args)`. For the functions like stack, where clip boundaries will be apparent because, well it's stacking the left border against the right border of a (assuming you're using the *comparison* functions to compare stuff here) similar frame, I left them as just `*args` since doing `stack_vertical(clipa, clipb, clipc)` is probably easier for a basic script writer. For cases where clip boundaries or the order might be unclear, I put them as `**kwargs` so something like `split(BD=bdmv, TV=src1)` could be done. If the user has them in a list `clips=[...]` already, `func(*clips)` will work fine in the `*args` functions. If the user wants to do something special with it, they can just initialize the classes directly with the additional params like `label_alignment`.

Comment (FichteFoll) [reply lvsfunc/comparison.py:593]: Not a fan of the inconsistency, personally.

Comment (OrangeChannel) [reply lvsfunc/comparison.py:593]: louis suggested a way to accept both types and error if both are given, which is something I'll look into.

Regarding kwargs here, it's fine because it puts them into a dict before sending them to the class constructor, so no keyword collision here.

**Comment by LightArrowsEXE** [comment]
Tested all but latest commit, and it seemed to work. PR looks good enough for merging to me. How about you, @heicrd?

**Comment by heicrd** [comment]
I haven't had time to review this but if @FichteFoll has OK'd it it should be good

**Comment by OrangeChannel** [comment]
I can still implement the “accepts args and kwargs” suggestion if you think that’d be a good idea for the shortcut functions at the bottom, or just completely get rid of the shortcut functions if you think using `Stack([clipa, clipb], direction=Direction.VERTICAL).clip` is simple enough for users compared to `stack_vertical(clipa, clipb)` or `Split({'clipa': srca, 'clipb': srcb}).clip` compared to `split(clipa=srca, clipb=srcb)`. If you’re alright with the way it currently works, I’m fine with the code how it is now though would be nice to have a way of testing these without using actual images in the repo. If someone can help me with that (I think someone has mentioned using makediff or expr) I can set up the tests today, just not sure how to compare the outputs nor where we want tests to go in the repo yet.

File: lvsfunc/comparison.py (hunk 23 above)
Comment (OrangeChannel) [reply lvsfunc/comparison.py]: Made it so all the functions accept both positional arg clips and named kwarg clips, and errors out if both types of args are given.

**Comment by OrangeChannel** [comment]
All functions/classes/documentation tested and working.
